test/office.server.test.js

```diff
@@ -0,0 +1,83 @@
+const assert = require("assert");
+const io = require("socket.io-client");
+const sinon = require("sinon");
+
+const app = require("../backend/app.server");
+const Office = require("../backend/office.server");
+
+const SOCKET_HOST = "0.0.0.0";
+const SOCKET_PORT = "3003";
+
+describe("Office server", () => {
+  const user = JSON.stringify({ id: 1 });
+  const room = JSON.stringify({});
+  const ioOptions = {
+    transports: ["websocket"],
+    forceNew: true,
+    reconnection: false,
+    query: `user=${user}&room=${room}`,
+  };
+  let officeController;
+  let server;
+  let office;
+  let socketClient;
+
+  beforeEach(() => {
+    officeController = {
+      addUserInRoom: sinon.fake(),
+      getUserInRoom: sinon.fake(),
+      getUsersInOfficeByMap: sinon.fake(),
+      removeUser: sinon.fake(),
+    };
+
+    // start express server
+    server = app.listen(SOCKET_PORT, SOCKET_HOST);
+
+    // start Office server
+    office = new Office(officeController, server);
+
+    // connect a client to Office server
+    socketClient = io.connect(`http://${SOCKET_HOST}:${SOCKET_PORT}/`, ioOptions);
+  });
+
+  afterEach(() => {
+    if (socketClient) {
+      socketClient.disconnect();
+    }
+
+    server.close();
+  });
+
+  it("should connect and disconnect with success", (done) => {
+    office.start();
+
+    socketClient.on("connect", () => {
+      assert(officeController.addUserInRoom.calledOnce);
+      assert(officeController.getUserInRoom.calledOnce);
+      assert(officeController.getUsersInOfficeByMap.calledOnce);
+      done();
+    });
+  });
+
+  it("should enter a room with success", (done) => {
+    const payload = {
+      room: "test-room",
+    };
+
+    office.start();
+
+    socketClient.on("connect", () => {
+      socketClient.on("enter-room", () => {
+        if (officeController.addUserInRoom.callCount === 2) {
```

Comment:
I honestly do not like too much this kind of logic inside a test, because IMHO it adds more conditions to set up the test and make it harder to reproduce and understand. 

Can we simplify this logic/test to be more clear?

Reply:
Thanks for bring up this topic. I spent a lot of time analyzing the implementation and how I could test it.
When the backend connect, it add the current user in a default room, this trigger the "enter-room" event to the client. So, when I emit the "enter-room" from the client, I must validate the second event "enter-room" from the backend.
This leads me to raise some questions about implementation:

- Could the user connect and stay in a neutral area, like a lobby or hall? It is a region where the user is logged in, but would make it clear that he is not in any meeting.
- It is necessary trigger the "enter-room" event when the user is connected?
- If the answer is negative for the two questions above, a cleaner solution would be to pass a callback as the second parameter, through it it would be easier to validate if the response was of the event emited.

Reply:
The default room when user connect is `room-1` https://github.com/ResultadosDigitais/matrix/blob/master/backend/office.server.js#L3

My suggestion to test this is, test both cases, when `connect` and when `enter-rom`. This will clarify the current behavior.

```
 socketClient.on("connect", () => { 
       assert.equal(
            officeController.addUserInRoom.getCall(0).lastArg,
            "room-1",
          );

 socketClient.on("enter-room", () => {
          assert.equal(
            officeController.addUserInRoom.getCall(1).lastArg,
            "test-room",
          );
          done();
...
```
Makes sense?

Reply:
I agree with @alesshh :)

Reply:
Great discussion, @alesshh @lccezinha . Besides improving the test, it's helping me to better understand the project.

The conditional is necessary because the event is triggered twice, so without the "if" the code validates only on the first trigger. I even include a comment explaining to avoid doubts of who is reading the test. If you guys prefer, I can switch to a triggered event counter instead of using the method call count. What do you think?

About the default room, I create **two more tests**. Analyzing the code, I saw that it uses the default room if a room is not set when connecting, so I've created a test for each of these two cases separately.

Let me know if these changes made sense.

Reply:
Congratz @megatroom the code seems a lot better right now!